client/lib/browser-storage/bypass.ts

```diff
@@ -11,7 +16,20 @@ const debug = debugModule( 'calypso:support-user' );
 // a user's data while support-user is active, ensuring it does not contaminate
 // the original user, and vice versa.
 
-const memoryStore = new Map< string, any >();
+const memoryStore = new Map();
+
+export async function getAllStoredItems( pattern?: RegExp ): Promise< StoredItems > {
+	debug( 'browser-storage bypass', 'getAllStoredItems' );
+
+	// TODO: Replace below with Object.fromEntries when it's available (needs core-js@3).
+	const results: StoredItems = {};
+	for ( const [ key, value ] of memoryStore.entries() ) {
+		if ( ! pattern || pattern?.test( key ) ) {
+			results[ key ] = value;
+		}
+	}
+	return results;
+}
```

Comment:
given the very generic nature of `StoredItems` I'm suspicious of its value in `types.ts`. I acknowledge that I may be pre-judging the name `StoredItems` based on an assumption that's it's generic.

we're also doing something interesting here which is decoupling the type of the actual storage from the type of the returned value. I think we have no idea what types of data we'll store here and so I can image value in either leaving the types as they were, defining `StoredItems` here in this file, making the type in `types.ts` more specific (such as naming it `JSONSerializable`), or going further to synchronize the types

```ts
// types.ts
type JSONValue =
  | string
  | number
  | boolean
  | null
  | Array<JSONValue>

type JSONSerializable = {
  [key: string]: JSONValue | JSONSerializable;
};
```

```ts
// bypass.ts
const memoryStore = new Map<string, JSONSerializable>();

export async function getAllStoredItems( pattern?: RegExp ): Promise< JSONSerializable > {
 	// TODO: Replace below with Object.fromEntries when it's available (needs core-js@3).
 	const results: JSONSerializable = {};
 	for ( const [ key, value ] of memoryStore.entries() ) {
 		if ( ! pattern || pattern?.test( key ) ) {
 			results[ key ] = value;
 		}
 	}
 	return results;
 }
```

Mostly musing here, but the use of `JSONSerializable` would prevent a number of possible cases trying to store objects with known types that cannot be serializable.

```ts
const j: JSONSerializable = {
  foo: () => { }, // Type '() => void' is not assignable to type 'string | number | boolean | JSONValue[] | JSONSerializable | null'.
  bar: new Map(), // Type 'Map<any, any>' is not assignable to type 'string | number | boolean | JSONValue[] | JSONSerializable | null'.
}
```

Reply:
IndexedDB can serialize and store much more than just the types that are allowed by JSON. It can do Maps, Sets, Blobs, Files, Dates and RegExps, ArrayBuffers and ImageData. Basically any value except functions, DOM elements and Symbols. See the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) docs for more details.

Is there a TypeScript type for all that other than `any`?

I had a quick look at the [Dexie](https://github.com/dfahlander/Dexie.js) library, a popular IDB wrapper. It's written in TypeScript and uses `any` for the object store values, too.